### ISSUE TITLE
better design

### ISSUE THREAD

**Issue by marklrh** [marklrh/otter#1]
It would be good to come up with a good design to solve this [problem](http://stackoverflow.com/questions/30269930/lwt-stream-usage-and-design-concern). 

@rgrinberg would you like to give my some suggestions? ivg's edit is not quite satisfying.

**Comment by rgrinberg** [comment]
Could you detail what's unsatisfying with @ivg's answer?

**Comment by marklrh** [comment]
Currently, we have
````OCaml
let get_follower ~access_token ~screen_name ?(count=5000) ?(wait=true) () =
      let base_uri = "https://api.twitter.com/1.1/followers/ids.json" in
      let cursor = ref (-1) in
      let f () =
        print_endline "api call";
        if !cursor = 0 then return None else
        Client.do_get_request
          ~uri_parameters:
            ["screen_name", screen_name;
             "cursor", (string_of_int (!cursor));
             "count", (string_of_int count)]
          ~uri:(Uri.of_string base_uri)
          ~access_token:access_token
        () >>= fun res ->
        match res with
        | `Ok (h, b_str) -> begin
          match (Yojson.Safe.from_string b_str |> user_ids_of_yojson) with
          | `Ok uids -> 
            cursor := uids.next_cursor;
            return (Some (`Ok uids))
          | `Error msg -> failwith msg 
          end
        | `Error e -> return (Some (`Error (process_error_exn e))) in
      Lwt_stream.from f
````

As Drup suggested, ref implementation is already good, but some further abstraction can be made
````OCaml
let state_update () = 
  let cursor = ref (-1) in
  let update uids = 
    cursor := uids.next_cursor in
  cursor, update

 let get_follower ~access_token ~screen_name ?(count=5000) ?(wait=true) () =
      let base_uri = "https://api.twitter.com/1.1/followers/ids.json" in
      let cursor, update = state_update () in
      let f u () =
        print_endline "api call";
        if !cursor = 0 then return None else
        Client.do_get_request
          ~uri_parameters:
            ["screen_name", screen_name;
             "cursor", (string_of_int (!cursor));
             "count", (string_of_int count)]
          ~uri:(Uri.of_string base_uri)
          ~access_token:access_token
        () >>= fun res ->
        match res with
        | `Ok (h, b_str) -> begin
          match (Yojson.Safe.from_string b_str |> user_ids_of_yojson) with
          | `Ok uids ->
             update uids; 
             return (Some (`Ok uids))
          | `Error msg -> failwith msg 
          end
        | `Error e -> return (Some (`Error (process_error_exn e))) in
      Lwt_stream.from (f update)
```` 

or we can simply have
````OCaml
let state_update () = 
  let cursor = ref (-1) in
  let update uids = 
    cursor := uids.next_cursor in
  cursor, update

 let get_follower ~access_token ~screen_name ?(count=5000) ?(wait=true) () =
      let base_uri = "https://api.twitter.com/1.1/followers/ids.json" in
      let cursor, update = state_update () in
      let f () =
        print_endline "api call";
        if !cursor = 0 then return None else
        Client.do_get_request
          ~uri_parameters:
            ["screen_name", screen_name;
             "cursor", (string_of_int (!cursor));
             "count", (string_of_int count)]
          ~uri:(Uri.of_string base_uri)
          ~access_token:access_token
        () >>= fun res ->
        match res with
        | `Ok (h, b_str) -> begin
          match (Yojson.Safe.from_string b_str |> user_ids_of_yojson) with
          | `Ok uids ->
             update uids; 
             return (Some (`Ok uids))
          | `Error msg -> failwith msg 
          end
        | `Error e -> return (Some (`Error (process_error_exn e))) in
      Lwt_stream.from f
````

**Comment by marklrh** [comment]
ivg's answer does not make API calls lazy. API calls are precious and I'll consider more about API call limit than performance consideration

**Comment by rgrinberg** [comment]
That's true, @ivs's answer does not make the API calls fully lazy. But I think what's even worse is that your API makes batching requests impossible!

A much more important optimization is to be able to batch requests and execute over some connection pool (for example)

**Comment by ivg** [comment]
@marklrh, you should be accurate with the terminology, or you can confuse yourself. Lwt is not lazy. A term of type `'a Lwt.t` is much different from a term of type `'a Lazy.t`. The first one just encapsulates the ongoing computation, that is already started but possibly not yet finished. The latter is a computation is either never started, or is already done. In other words all `'a Lwt.t` are things that are already ongoing, or will become ongoing as soon as it is possible. That's why Lwt threads are eager by definition.

What concerning your problem, you can use `Lwt_stream.create_bounded` that will reduce the amount of prefetching calls. For example, with the length of the queue equal to `0` it will just act as a mailbox and will do only one call to the server per invocation of the `get_follower` function. E.g., 
```ocaml
  let stream,q = Lwt_stream.create_bounded batch_length in
  let rec loop cursor () = 
    make_request cursor >|= parse_response >>= function
    | Ok {next_cursor=0} -> return q#close
    | Ok {next_cursor=n} as r -> q#push r >>= loop cursor
    | error -> q#push error >>= loop cursor in
  async (loop (-1));
  stream
```

**Comment by marklrh** [comment]
yeah I mean ``Lwt_stream.t`` is lazy.

You mentioned on StackOverlfow that 
> But you will still have some amount of prefetching (or back to our pipe system analogy, you will have some kind of extension tank)

well, I think any amount of prefetching is not OK. Some API calls are limited to 15 times in a 15 minutes window, which means that API calls are precious and could not be wasted.

**Comment by marklrh** [comment]
as for the batch request, it is not quite possible.

For now, we have "``request1`` -> ``response1`` -> ``request2`` -> ``response2`` -> ..."

The reason we cannot do batch is that the value of "cursor" in ``request2`` depends on the value of "next_cursor" in ``response1``.

**Comment by marklrh** [comment]
hi @ivg

I did some test with ``Lwt_stream.create_bounded`` and ``async (loop -1)``. I noticed that when async starts (the same time get_follower is used to create a stream), it always call a push function first, no matter if the user is going to use stream or not. I don't know if this behavior is negligible.

**Comment by ivg** [comment]
Yes, this is the tank. It will prefetch (or pump) up to the `batch_length` followers. By prefetching I mean, that it will resolve this number of followers and put them in a queue. As user removes followers from the queue, it will be filled back. At maximum possible rate. With `batch_length` equal to `0` or `1`  the queue will be actually a mbox variable.

**Comment by marklrh** [comment]
I don't know if ``created_bounded`` is ok because if user create a stream but he does not use it, he will use one API call. However, this case is nearly impossible to happen...

**Comment by ivg** [comment]
Well, if you're interested in mine opinion, then I wouldn't use `Lwt_stream` at all, or at least expose it to the user. First of all, stream leaks the abstraction (in particular it exposes, that followers are sequence, not a set), second, stream is exhaustible and thus, very error-prone. Moreover, as soon as you expose a stream interface it would be hard for you to change implementation later. 

Since followers are conceptually a set, not an ordered sequence, I will build my interface around the set of abstract followers. But followers can't be first class, since we have this cursor, that means, that we need to store this information in some context. Ideally, we can use [Environment Comonad][1] here, or just pass the environment around. You will still need to use mutable fields here, for transferring this hidden cursor information from client to client. But with a context, you will have a freedom to change everything later. So, it may looks something like this:

```ocaml

(* a context *)
type user

(* each follower captures a context inside *)
type follower 

val followers : user -> follower Sequence.t

module Follower : sig
    type t = follower
    val name : follower -> string Lwt.t
    ...
end

```

[1]: https://hackage.haskell.org/package/comonad-4.2.6/docs/Control-Comonad-Env.html
